lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -840,66 +844,73 @@ instance W.KeyToAddress t Seq.SeqKey => PersistState (Seq.SeqState t) where
                 (uncurry (==))
         let eGap = Seq.gap extPool
         let iGap = Seq.gap intPool
-        ssid <- insert (SeqState wid sl eGap iGap (AddressPoolXPub xpub))
-        insertAddressPool ssid intPool
-        insertAddressPool ssid extPool
-        insertMany_ $ mkSeqStatePendingIxs ssid $ Seq.pendingChangeIxs st
+        repsert (SeqStateKey wid) (SeqState wid eGap iGap (AddressPoolXPub xpub))
+        insertAddressPool wid sl intPool
+        insertAddressPool wid sl extPool
+        cpid <- insert (SeqStateCheckpoint wid sl)
+        insertMany_ $ mkSeqStatePendingIxs cpid $ Seq.pendingChangeIxs st
 
     selectState (wid, sl) = runMaybeT $ do
-        st <- MaybeT $ selectFirst
-            [ SeqStateWalletId ==. wid
+        st <- MaybeT $ selectFirst [SeqStateWalletId ==. wid] []
+        stc <- MaybeT $ selectFirst
+            [ SeqStateCheckpointWalletId ==. wid
             , SeqStateCheckpointSlot ==. sl
             ] []
-        let (ssid, SeqState _ _ eGap iGap xPub) = (entityKey st, entityVal st)
-        intPool <- lift $ selectAddressPool ssid iGap xPub
-        extPool <- lift $ selectAddressPool ssid eGap xPub
-        pendingChangeIxs <- lift $ selectSeqStatePendingIxs ssid
+
+        let SeqState _ eGap iGap xPub = entityVal st
+        intPool <- lift $ selectAddressPool wid sl iGap xPub
+        extPool <- lift $ selectAddressPool wid sl eGap xPub
+        pendingChangeIxs <- lift $ selectSeqStatePendingIxs (entityKey stc)
         pure $ Seq.SeqState intPool extPool pendingChangeIxs
 
     deleteState wid = do
-        ssid <- fmap entityKey <$> selectList [ SeqStateWalletId ==. wid ] []
-        deleteWhere [ SeqStateAddressesSeqStateId <-. ssid ]
-        deleteWhere [ SeqStatePendingIxSeqStateId <-. ssid ]
+        cpid <- selectList [ SeqStateCheckpointWalletId ==. wid ] []
+        deleteWhere [ SeqStatePendingIxCheckpointId <-. fmap entityKey cpid ]
```

Comment:
Not _directly_ relevant to this, but these are run as "separate SQL statements"?

Should we use something like [Sqlite transactions](https://www.tutlane.com/tutorial/sqlite/sqlite-transactions-begin-commit-rollback), to protect against inconsistent state from the program crashing in the middle of a complex sequence of deletes?

We do use `DB.withLock` from the [`WalletLayer`](https://github.com/input-output-hk/cardano-wallet/blob/54ab0a01d02d2fb5ca40365bc3746d6e4c60ecb6/lib/core/src/Cardano/Wallet.hs#L581), but I think that just protects against multiple wallets/requests mutating the DB at the same time.

Reply:
Indeed (as mentioned here: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet.hs#L682-L685). The lock is only there to prevent concurrent issues in case clients are making concurrent conflicting requests, but it doesn't prevent data corruption / inconsistency like a transaction would do. 

I recall @rvl had a reason for not using sqlite transaction here. I think it was something about being only partially supported by our driver library :thinking: ?

Reply:
Very good point.
It _is_ running each DBLayer method in a transaction because we are using [runSqlConn](http://hackage.haskell.org/package/persistent-2.9.2/docs/Database-Persist-Sql.html#v:runSqlConn).
At the time of writing the initial DBLayer we were a bit unsure where to draw the lines with transactions and locking.
We ought to have another look. In particular, see if we can put a transaction around `restoreBlocks` in the wallet layer.